pagefaultgames/pokerogue
# [Item] Add Scope Lens and Leak

## What are the changes?


Completely adds and implements the following items: [`Scope Lens`](https://bulbapedia.bulbagarden.net/wiki/Scope_Lens) and [`Leek`](https://bulbapedia.bulbagarden.net/wiki/Leek).

With the addition of `Leek`using its intended sprite, `Baton`'s sprite was also changed.

## Why am I doing these changes?




As usual, this PR mainly aims to continue the incorporation of useful items that are not yet included in the game from the mainline titles. Although not directly asked for, `Leek` essentially increases the meta-viability of the Farfetch'd line, similar to the items introduced in #1899. However, `Scope Lens` in particular has been widely requested in general for its utility and for the purpose of making critical hits more reliable in endless modes.

## What did change?



The item internal functionality was achieved by creating two new classes, `CritBoosterModifier` extending `PokemonHeldItemModifier` and `SpeciesCritBoosterModifierr` extending `CritBoosterModifier` in `src/modifier/modifier.ts` and adding a line to `Pokemon`'s `apply()` function in `src/field/pokemon.ts` to provide the critical-hit stage boost prior to the determination of a critical hit.

`Scope Lens` was given a weight of `4` in the Rogue tier since it can be used with any `Pokemon`. A weight function was provided for `Leek` provides a weight of `12` if a `Pokemon` from the Farfetch'd line is in the party and does not already have `Leek`; otherwise, a weight of `0` is given, preventing the item from being pushed into the reward pool when it has no utility with the current team.

Localization was provided for `de`, `en`, `es`, `fr`, `it`, `ko`, `zh_CN`, and `zh_TW` directly from Scarlet/Violet using the following resources:
- `Scope Lens` - [name](https://abcboy101.github.io/poke-corpus/#id=sv.itemname.ITEMNAME_232), [description](https://abcboy101.github.io/poke-corpus/#id=sv.iteminfo.ITEMINFO_232)
- `Leek` - [name](https://abcboy101.github.io/poke-corpus/#id=sv.itemname.ITEMNAME_259), [description](https://abcboy101.github.io/poke-corpus/#id=sv.iteminfo.ITEMINFO_259)

All translations were taken mostly verbatim, taking out superfluous text as needed with some educated guessing.

The new `Baton` sprite was included as an individual `.png` file, and TexturePacker was used along with the repo's `configuration.tps` file to generate a newly optimized texture atlas (`items.png`) along with an updated `items.json` file. Note that `Leek` and `Scope Lens` already had sprites. To accommodate the new sprite and the items making use of existing sprites, the `Leek` sprite was renamed to `leek.png` from `stick.png`, and `Baton`'s entry in `src/modifier/modifier-type.ts` was changed to use the new `baton` image instead of `stick`. The `Baton` sprite itself was provided by Diashi from the art team.

Unit tests were included for both items as well to stress test the items and ensure they function as intended.

Lastly, a `console.log` line was added to make it easy to see what the current critical-hit stage is at the time of a hit (using the vernacular from Bulbapedia's [critical hit page](https://bulbapedia.bulbagarden.net/wiki/Critical_hit#Stage_modifiers).

### Screenshots/Videos



#### Appearance, Name, and Description (`en`)

| <img width="657" alt="scopeLensAppearance" src="https://github.com/pagefaultgames/pokerogue/assets/109637146/b1aa5ca8-8fcb-4116-a4d2-344162b15773"> | <img width="657" alt="leekAppearance" src="https://github.com/pagefaultgames/pokerogue/assets/109637146/c06174f0-9713-4306-a4e1-3e7cc0bffef8"> |
| :---: | :---: |
| `Scope Lens` | `Leek` |

Note `Baton`'s appearance change in both images. 

#### Application - No Critical-Hit Chance Boost

<img width="439" alt="critStage0" src="https://github.com/pagefaultgames/pokerogue/assets/109637146/22fd4511-abf6-44bc-bc9c-5f31cad354ca">

#### Application - `Scope Lens`

<img width="437" alt="scopeLensApplied" src="https://github.com/pagefaultgames/pokerogue/assets/109637146/8b7f4f54-ea33-40f8-81a9-647c80c44887">

#### Application - `Leek` (on Farfetch'd)

<img width="437" alt="leekApplied" src="https://github.com/pagefaultgames/pokerogue/assets/109637146/f36125d3-7855-4437-bdad-2ed03001d4a3">

## How to test the changes?





For general testing related to item functionality, the `src/override.ts` file was modified such that the `STARTING_HELD_ITEMS_OVERRIDE` array included both or one of the items.
```ts
export const STARTING_HELD_ITEMS_OVERRIDE: Array<ModifierOverride> = [{ name: "SCOPE_LENS" }, { name: "LEEK" }];
```

`ITEM_REWARD_OVERRIDE` can also be used to see the item descriptions more clearly along with the new `Baton` sprite:
```ts
export const ITEM_REWARD_OVERRIDE: Array<String> = [ "SCOPE_LENS", "LEEK", "BATON" ];
```

`STARTER_SPECIES_OVERRIDE` was also used to test if the Farfetch'd line was affected by `Leek`: `Species.FARFETCHD`, `Species.GALAR_FARFETCHD`, and `Species.SIRFETCHD`. Any `Species` can be used with `Scope Lens`.

Additionally, the different translations can be seen by cycling through the language settings.

## Checklist
